dotcom-rendering/src/lib/useMatchMedia.ts

```diff
@@ -0,0 +1,39 @@
+import { useSyncExternalStore } from 'react';
+
+/**
+ * Stores constructed media queries by their query string, so they don't have
+ * to be recreated on each render. Also allows multiple components to use the
+ * same media query.
+ */
+const cachedQueries = new Map<string, MediaQueryList>();
+
+/**
+ * A custom hook to test the provided media query. Will return a `boolean`
+ * signifying whether the media query currently matches, and will trigger a
+ * re-render whenever the match status changes in either direction;
+ * e.g. from `true` to `false`, or from `false` to `true`.
+ *
+ * @param query A media query string for use with `window.matchMedia`
+ * @returns A `boolean` signifying whether the media query currently matches
+ */
+const useMatchMedia = (query: string): boolean => {
+	const maybeQuery = cachedQueries.get(query);
+	const mediaQuery = maybeQuery ?? window.matchMedia(query);
+
+	if (maybeQuery === undefined) {
+		cachedQueries.set(query, mediaQuery);
+	}
```

Comment:
Non-blocking suggestion. Depends on your appetite for one liners but ymmv:

```suggestion
	const mediaQuery = cachedQueries.get(query) ??
	  cachedQueries.set(query, window.matchMedia(query)).get(query);
```

Reply:
Unfortunately if you do this you get a type error because `.get` returns `A | undefined` 😔.